File: examples/PaymentService.java
package com.example.payment;

import com.example.payment.exception.PaymentException;
import com.example.payment.model.Transaction;
import com.example.payment.model.CardDetails;
import com.example.payment.repository.TransactionRepository;
import com.example.payment.gateway.PaymentGatewayClient;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.UUID;

@Service
public class PaymentService {

    private final TransactionRepository transactionRepository;
    private final PaymentGatewayClient gatewayClient;

    @Autowired
    public PaymentService(
        TransactionRepository transactionRepository,
        PaymentGatewayClient gatewayClient
    ) {
        this.transactionRepository = transactionRepository;
        this.gatewayClient = gatewayClient;
    }

    public Transaction processPayment(CardDetails card, double amount, String currency)
        throws PaymentException {
        if (amount <= 0) {
            throw new PaymentException("Amount must be positive");
        }
        validateCard(card);
        String gatewayRef = gatewayClient.charge(card.getToken(), amount, currency);
        Transaction txn = new Transaction(UUID.randomUUID().toString(), amount, currency, "COMPLETED", gatewayRef);
        return transactionRepository.save(txn);
    }

    public Transaction refund(String transactionId, double amount) throws PaymentException {
        Transaction txn = transactionRepository.findById(transactionId)
            .orElseThrow(() -> new PaymentException("Transaction not found: " + transactionId));
        if (amount > txn.getAmount()) {
            throw new PaymentException("Refund exceeds original amount");
        }
        gatewayClient.refund(txn.getGatewayRef(), amount);
        txn.setStatus("REFUNDED");
        return transactionRepository.save(txn);
    }

    public Optional<Transaction> getTransaction(String transactionId) {
        return transactionRepository.findById(transactionId);
    }

    public boolean validateCard(CardDetails card) throws PaymentException {
        if (card == null || card.getToken() == null || card.getToken().isBlank()) {
            throw new PaymentException("Invalid card token");
        }
        return true;
    }
}

File: examples/payment_service.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Optional


@dataclass
class CardDetails:
    token: str
    last_four: str
    expiry: str


@dataclass
class Transaction:
    id: str
    amount: float
    currency: str
    status: str
    created_at: datetime


class PaymentService:
    """Core payment processing service."""

    def __init__(self, db, gateway_client) -> None:
        self._db = db
        self._gateway = gateway_client

    def charge(self, card: CardDetails, amount: float, currency: str = "USD") -> Transaction:
        """Charge a card and persist the transaction."""
        if not self.validate_card(card):
            raise ValueError("Invalid card details")
        if amount <= 0:
            raise ValueError("Amount must be positive")

        result = self._gateway.charge(card.token, amount, currency)
        txn = Transaction(
            id=result["id"],
            amount=amount,
            currency=currency,
            status="completed",
            created_at=datetime.utcnow(),
        )
        self._db.save(txn)
        return txn

    def refund(self, transaction_id: str, amount: Optional[float] = None) -> bool:
        """Refund a transaction fully or partially."""
        txn = self._db.get(transaction_id)
        if txn is None:
            raise ValueError(f"Transaction not found: {transaction_id}")
        refund_amount = amount or txn.amount
        if refund_amount > txn.amount:
            raise ValueError("Refund amount exceeds original charge")
        return self._gateway.refund(transaction_id, refund_amount)

    def get_transaction(self, transaction_id: str) -> Optional[Transaction]:
        """Fetch a transaction by ID. Returns None if not found."""
        return self._db.get(transaction_id)

    def validate_card(self, card: CardDetails) -> bool:
        """Validate card token is non-empty and expiry is in the future."""
        if not card.token or not card.expiry:
            return False
        try:
            expiry = datetime.strptime(card.expiry, "%m/%y")
            return expiry > datetime.utcnow()
        except ValueError:
            return False
